dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/aggregation/AggregationPipelineQueryNode.java

```diff
@@ -0,0 +1,882 @@
+        Bson projectOpts = Projections.fields(
+                Projections.computed(VALUES, Projections.fields(valueFields)),
+                Projections.computed(HASHES, Projections.fields(hashFields)),
+                Projections.computed(TYPES, Projections.fields(typeFields)),
+                Projections.include(LEVEL),
+                Projections.include(TIMESTAMP));
+        pipeline.add(Aggregates.project(projectOpts));
+        return true;
+    }
+
+    /**
+     * Add a SPARQL filter to the pipeline, if possible. A filter eliminates
+     * results that don't satisfy a given condition. Not all conditional
+     * expressions are supported. If unsupported expressions are used in the
+     * filter, the pipeline will remain unchanged and this method will return
+     * false.
+     * @param condition The filter condition
+     * @return True if the filter was successfully converted into a pipeline
+     *  step, false otherwise.
+     */
+    public boolean filter(ValueExpr condition) {
+        if (condition instanceof Compare) {
+            Compare compare = (Compare) condition;
+            Compare.CompareOp operator = compare.getOperator();
+            Object leftArg = valueFieldExpr(compare.getLeftArg());
+            Object rightArg = valueFieldExpr(compare.getRightArg());
+            if (leftArg == null || rightArg == null) {
+                // unsupported value expression, can't convert filter
+                return false;
+            }
+            final String opFunc;
+            switch (operator) {
+            case EQ:
+                opFunc = "$eq";
+                break;
+            case NE:
+                opFunc = "$ne";
+                break;
+            case LT:
+                opFunc = "$lt";
+                break;
+            case LE:
+                opFunc = "$le";
+                break;
+            case GT:
+                opFunc = "$gt";
+                break;
+            case GE:
+                opFunc = "$ge";
+                break;
+            default:
+                // unrecognized comparison operator, can't convert filter
+                return false;
+            }
+            Document compareDoc = new Document(opFunc, Arrays.asList(leftArg, rightArg));
+            pipeline.add(Aggregates.project(Projections.fields(
+                    Projections.computed("FILTER", compareDoc),
+                    Projections.include(VALUES, HASHES, TYPES, LEVEL, TIMESTAMP))));
+            pipeline.add(Aggregates.match(new Document("FILTER", true)));
+            pipeline.add(Aggregates.project(Projections.fields(
+                    Projections.include(VALUES, HASHES, TYPES, LEVEL, TIMESTAMP))));
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Add a $group step to filter out redundant solutions.
+     * @return True if the distinct operation was successfully appended.
+     */
+    public boolean distinct() {
```

Comment:
there is a mongo function for finding distinct values over fields

Reply:
yeah I misunderstood, I though SPARQL DISTINCT was similar to mongo's distinct, or vis versa.  I mean you might be able to use distinct, then do another operation to split up the resulting field into separate documents, but if that's the case this is fine.